connector/src/main/java/org/jboss/as/connector/subsystems/datasources/GetDataSourceClassInfoOperationHandler.java

```diff
@@ -0,0 +1,235 @@
+ */
+@SuppressWarnings("deprecation")
+// ModuleIdentifier is used in InstalledDriver, which is deprecated.
+public class GetDataSourceClassInfoOperationHandler implements OperationStepHandler {
+
+    public static final GetDataSourceClassInfoOperationHandler INSTANCE = new GetDataSourceClassInfoOperationHandler();
+
+    private GetDataSourceClassInfoOperationHandler() {
+    }
+
+    @Override
+    public void execute(final OperationContext context, final ModelNode operation) throws OperationFailedException {
+
+        final String driverName = context.getCurrentAddressValue();
+        if (context.isNormalServer()) {
+            context.addStep(new OperationStepHandler() {
+
+                @Override
+                public void execute(final OperationContext context, final ModelNode operation) throws OperationFailedException {
+
+                    ServiceController<?> sc = context.getServiceRegistry(false).getRequiredService(
+                            ConnectorServices.JDBC_DRIVER_REGISTRY_SERVICE);
+                    DriverRegistry driverRegistry = DriverRegistry.class.cast(sc.getValue());
+                    InstalledDriver driver = driverRegistry.getInstalledDriver(driverName);
+                    if (driver == null) {
+                        context.getResult().set(new ModelNode());
+                        return;
+                    }
+                    ModelNode result = dsClsInfoNode(driver.getModuleName(), driver.getDataSourceClassName(),
+                            driver.getXaDataSourceClassName());
+                    context.getResult().set(result);
+                }
+
+            }, OperationContext.Stage.RUNTIME);
+        }
+    }
+
+    static ModelNode dsClsInfoNode(ModuleIdentifier mid, String dsClsName, String xaDSClsName) throws OperationFailedException {
+        ModelNode result = new ModelNode();
+        if (dsClsName != null) {
+            ModelNode dsNode = new ModelNode();
+            dsNode.get(dsClsName).set(findPropsFromCls(mid, dsClsName));
+            result.add(dsNode);
+        }
+        if (xaDSClsName != null) {
+            ModelNode xaDSNode = new ModelNode();
+            xaDSNode.get(xaDSClsName).set(findPropsFromCls(mid, xaDSClsName));
+            result.add(xaDSNode);
+        }
+        return result;
+    }
+
+    private static ModelNode findPropsFromCls(ModuleIdentifier mid, String clsName) throws OperationFailedException {
+        final ModelNode result = new ModelNode();
+        findPropsMap(mid, clsName).forEach((k, v) -> {
+            result.get(k).set(v.getTypeName());
+        });
+        return result;
+    }
+
+    private static Map<String, Type> findPropsMap(ModuleIdentifier mid, String clsName) throws OperationFailedException {
+        Class<?> cls = null;
+        ClassLoader cl = null;
+        if (mid != null) {
+            try {
+                cl = Module.getCallerModuleLoader().loadModule(mid).getClassLoader();
```

Comment:
Is this intended to work with drivers that are part of a deployment? If so then I think you may need to use ServiceModuleLoader here.

Reply:
JDBC driver that comes from a deployment won't be available in `/subsystem=datasources/jdbc-driver=XXX`, that maybe an issue to fix or an enhancement to implement. 

I think your suggestion is better, so I will switch to `ServiceModuleLoader`. thanks :)

Reply:
for deployed drivers server doesn't know datasource class, so it can't work anyway